File: mixin-base/src/main/java/net/yakclient/mixin/base/internal/loader/PackageTarget.java
package net.yakclient.mixin.base.internal.loader;

import java.util.Arrays;
import java.util.StringJoiner;

public class PackageTarget {
    protected final String[] path;

    public PackageTarget(String path) {
        this.path = fromPath(path);
    }

    public PackageTarget(String[] path) {
        this.path = path;
    }

    public static PackageTarget of(String path) {
        return new PackageTarget(fromPath(path));
    }

    public static PackageTarget of(Class<?> cls) {
        return new PackageTarget(fromPath(cls.getPackage().getName()));
    }


    protected static String[] fromPath(String path) {
        return path.split("\\.");
    }

    /**
     * Decides if the current path is equally or less
     * specific than then one given.
     *
     * @param target The path to compare against.
     * @return if the given path could be a child of the current target.
     */
    public boolean isTargetOf(PackageTarget target) {
        if (this.equals(target)) return true;

        boolean isTarget = false;
        for (int i = 0; i < this.path.length; i++) {
            isTarget = target.path.length > i && this.path[i].equals(target.path[i]);
        }

        return isTarget;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        final var that = (PackageTarget) o;
        return Arrays.equals(path, that.path);
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(path);
    }

    @Override
    public String toString() {
        final var joiner = new StringJoiner(".");
        for (var s : this.path) joiner.add(s);
        return joiner.toString();
    }
}

File: mixin-base/src/main/kotlin/net/yakclient/mixin/base/internal/loader/PackageTarget.kt
package net.yakclient.mixin.base.internal.loader

import java.util.*

open class PackageTarget(protected val path: List<String>) {
//    @JvmField
//    protected val path: Array<String>

//    constructor(path: String) {
//        this.path = fromPath(path)
//    }
//
//    constructor(path: Array<String>) {
//        this.path = path
//    }

    /**
     * Decides if the current path is equally or less
     * specific than then one given.
     *
     * @param target The path to compare against.
     * @return if the given path could be a child of the current target.
     */
    fun isTargetOf(target: PackageTarget): Boolean {
        if (this == target) return true
        var isTarget = false
        for (i in path.indices) {
            isTarget = target.path.size > i && path[i] == target.path[i]
        }
        return isTarget
    }

    override fun toString(): String {
        val joiner = StringJoiner(".")
        for (s in path) joiner.add(s)
        return joiner.toString()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as PackageTarget

        if (path != other.path) return false

        return true
    }

    override fun hashCode(): Int {
        return path.hashCode()
    }

    companion object {
        @JvmStatic
        fun of(path: String): PackageTarget {
            return PackageTarget(fromPath(path))
        }

        @JvmStatic
        fun of(cls: Class<*>): PackageTarget {
            return PackageTarget(fromPath(cls.getPackage().name))
        }

        @JvmStatic
        protected fun fromPath(path: String): List<String> {
            return path.split(".")
        }
    }
}